extensions/interactions/AlgebraicExpressionInput/AlgebraicExpressionInput.py

```diff
@@ -53,3 +54,242 @@ class AlgebraicExpressionInput(base.BaseInteraction):
+    @classmethod
+    def _to_answer_groups_proto(cls, answer_groups):
+        """Creates a AnswerGroup proto object
+        for AlgebraicExpressionInputInstanceDto.
+
+        Args:
+            answer_groups: list(AnswerGroup). List of answer groups of the
+                interaction instance.
+
+        Returns:
+            list. The AnswerGroup proto object list.
+        """
+        answer_group_list_proto = []
+        for answer_group in answer_groups:
+            base_answer_group_proto = answer_group.to_proto()
+            rules_spec_proto = cls._to_rule_specs_proto(answer_group.rule_specs)
+            answer_group_list_proto.append(
+                state_pb2.AlgebraicExpressionInputInstanceDto.AnswerGroupDto(
+                    base_answer_group=base_answer_group_proto,
+                    rule_specs=rules_spec_proto
+                )
+            )
+
+        return answer_group_list_proto
+
+    @classmethod
+    def _to_rule_specs_proto(cls, rule_specs_list):
+        """Creates a RuleSpec proto object.
+
+        Args:
+            rule_specs_list: list(RuleSpec). List of rule specifications.
+
+        Returns:
+            list. The RuleSpec proto object list.
+        """
+        rule_specs_list_proto = []
+
+        rule_type_to_proto_func_mapping = {
+            'MatchesExactlyWith': cls._to_matches_exactly_to_proto,
+            'IsEquivalentTo': cls._to_is_equivalent_to_proto,
+            'ContainsSomeOf': cls._to_contains_some_proto,
+            'OmitsSomeOf': cls._to_omit_some_proto,
+            'MatchesWithGeneralForm': cls._to_matches_with_form_proto
+        }
+
+        rule_type_to_proto_mapping = {
+            'MatchesExactlyWith': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    matches_exactly_with=x)),
+            'IsEquivalentTo': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    is_equivalent_to=x)),
+            'ContainsSomeOf': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    contains_some_of=x)),
+            'OmitsSomeOf': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    omits_some_of=x)),
+            'MatchesWithGeneralForm': lambda x: (
+                state_pb2.AlgebraicExpressionInputInstanceDto.RuleSpecDto(
+                    matches_with_general_form=x))
+        }
+
+        for rule_spec in rule_specs_list:
+            rule_type = rule_spec.rule_type
+            rule_proto = (
+                rule_type_to_proto_func_mapping[rule_type](
+                    rule_spec.inputs
+                )
+            )
+            rule_specs_list_proto.append(
+                rule_type_to_proto_mapping[rule_type](rule_proto)
+            )
+
+        return rule_specs_list_proto
+
+    @classmethod
+    def _to_matches_exactly_to_proto(cls, inputs):
```

Comment:
cls.convert_matches_exactly_rule_spec_to_proto()

(and similarly elsewhere)

Reply:
Fixed